lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp

```diff
@@ -119,6 +273,53 @@ void AnnotationSet::addAnnotations(ArrayAttr newAnnotations) {
   annotations = ArrayAttr::get(getContext(), annotationVec);
 }
 
+/// Remove an annotation from this annotation set. Returns true if any were
+/// removed, false otherwise.
+bool AnnotationSet::removeAnnotation(Annotation anno) {
+  return removeAnnotations([&](Annotation other) { return other == anno; });
+}
+
+/// Remove an annotation from this annotation set. Returns true if any were
+/// removed, false otherwise.
+bool AnnotationSet::removeAnnotation(Attribute anno) {
+  return removeAnnotations(
+      [&](Annotation other) { return other.getDict() == anno; });
+}
+
+/// Remove all annotations from this annotation set for which `predicate`
+/// returns true.
+bool AnnotationSet::removeAnnotations(
+    llvm::function_ref<bool(Annotation)> predicate) {
+  // Fast path for empty sets.
+  auto attr = getArrayAttr();
+  if (!attr)
+    return false;
+
+  // Search for the first match.
+  ArrayRef<Attribute> annos = getArrayAttr().getValue();
+  auto it = annos.begin();
+  while (it != annos.end() &&
+         !predicate(Annotation(it->cast<DictionaryAttr>())))
+    ++it;
+
+  // Fast path for sets where the predicate never matched.
+  if (it == annos.end())
+    return false;
+
+  // Build a filtered list of annotations.
+  SmallVector<Attribute, 8> filteredAnnos;
+  filteredAnnos.reserve(annos.size());
+  filteredAnnos.append(annos.begin(), it);
+  ++it;
+  while (it != annos.end()) {
+    if (!predicate(Annotation(it->cast<DictionaryAttr>())))
+      filteredAnnos.push_back(*it);
+    ++it;
+  }
```

Comment:
I'd personally just copy the attributes into a SmallVector, then use the 
` out.erase(std::remove_if(begin(out), end(out), std::not_fn(p)), end(out));`

idiom.  YMMV